design/annotation-for-power-cycling-and-deleting-failed-nodes.md

```diff
@@ -95,35 +95,51 @@ itself, allowing it to host workloads.
 
 ## Proposal
 
-This proposal calls for a new controller which watches for the presence of a
-specific Node annotation.  If present, the controller will locate the Machine
-and BareMetalHost host objects via their annotations, and (in serial) use the
-BareMetalHost API to power off the machine, delete the Node, and then power the
-machine back on again.
-
-In order to track the state of the recovery workflow, a [new
-CRD](https://github.com/kubevirt/machine-remediation/blob/master/pkg/apis/machineremediation/v1alpha1/machineremediation_types.go)
-is proposed.
+This proposal calls for a new machine remediation controller (MRC) in CAPBM
+which watches for the presence of a `host.metal3.io/external-remediation` Machine
+annotation. If present, the controller will locate the Machine
+and BareMetalHost host objects via their annotations, and will annotate the BareMetalHost
+CR with a `reboot.metal3.io/machine-remediation`. Upon host power off, the controller
+will delete the node and then remove the annotation, to power on the host by
+utilizing [Baremetal-Operator reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/reboot-interface.md).
+
+After restart/crash of the new controller it will check for
+`reboot.metal3.io/machine-remediation` annotation on all hosts,
+and continue remediation process, to make sure that such host is
+not left in powered off state.
+
+## Assumptions
+
+MHC checks that each machine has a node, and if it doesn't after some timeout,
+it will mark the machine as unhealthy. There could be situations where this timeout
+expires few seconds after MRC powered on the machine, resulting in one or more
+redundant reboot(s).
+Current MHC implementation checks that `now() - Machine.LastUpdated < timeout`.
+MRC deletes the unhealthy node, which updates `Machine.LastUpdated`.
+We assume that this implementation won't change.
+
+In addition, MRC has some operations to perform between node deletion (which triggers
+update of `Machine.LastUpdated`) to powering on the host, and these operations
+are also counted against that timeout. These operations are removing
+MAO's annotation and removing the reboot annotation. We assume that this
```

Comment:
There is no machine-api-operator upstream here in metal3. Is that going to change the timing at all?

Reply:
Since there's no machine-api-operator in metal3, no one will ever declare the machine is unhealthy (machine-healthcheck-controller [MHC] is part of MAO).
Somehow, MHC functionality should be added to metal3, and that metal3-MHC will have to consider these assumptions.

Reply:
> If Cluster API MachineHealthCheck is enough for us we might not need new healthcheck controller at all. We could just set annotations to Metal3Machines and/or BareMetalHost objects based on changes in Machine object.

The Cluster API MachineHealthCheck came from the same folks that wrote the one in machine-api-operator.  So my expectation is that they would (or could) be compatible.